server/wal/wal_rw_segment.go

```diff
@@ -154,6 +154,9 @@ func (ms *readWriteSegment) Append(offset int64, data []byte) error {
 	if len(data) == 0 {
 		return codec.ErrEmptyPayload
 	}
+	if !ms.HasSpace(len(data)) {
```

Comment:
It's an extra cost for validation, and we should avoid using it without `HasSpace` check. :) 

On the other hand, the panic is enough for correctness.

Reply:
cc @merlimat

Reply:
The `HasSpace()` is already checked with the WAL lock taken. I don't know how this condition could happen here. 

https://github.com/streamnative/oxia/blob/86cbb6a182c16198a69b7ec408aa04855eea27e7/server/wal/wal_impl.go#L287

Reply:
> The `HasSpace()` is already checked with the WAL lock taken. I don't know how this condition could happen here.
> 
> https://github.com/streamnative/oxia/blob/86cbb6a182c16198a69b7ec408aa04855eea27e7/server/wal/wal_impl.go#L287

1. This check should be in Segment append function, Because this object is the one that holds the length, and it can also prevent the situation where no inspection is carried out due to the failure of the caller.
2. panic maybe cause entire program exit, is this an elegant way to exit？

Reply:
> This check should be in Segment append function, Because this object is the one that holds the length, and it can also prevent the situation where no inspection is carried out due to the failure of the caller.

Yes, we could change the logic and return `errSegmentFull` on the append and avoid checking `HasSpace()` from outside.

> panic maybe cause entire program exit, is this an elegant way to exit？

No, though `panic()` is a good measure to protect against unexpected conditions. 

In this case I'm still not sure how the condition happens. If we check the segment size before, and we hold a lock on the whole WAL, what is the sequence of events that right now can lead to panic?

Reply:
Currently their is no sequence will cause append to segment exceeded size using WAL interface only, i just found it when writing unittest directly using segment struct.

I have move the logic only in segment.append function, PTAL